File: evals/refs/hull.py
import sys
def cross(o, a, b): return (a[0]-o[0])*(b[1]-o[1]) - (a[1]-o[1])*(b[0]-o[0])
def hull(points):
    pts = sorted(set(points))
    if len(pts) <= 1: return pts
    lower = []
    for p in pts:
        while len(lower) >= 2 and cross(lower[-2], lower[-1], p) <= 0: lower.pop()
        lower.append(p)
    upper = []
    for p in reversed(pts):
        while len(upper) >= 2 and cross(upper[-2], upper[-1], p) <= 0: upper.pop()
        upper.append(p)
    return lower[:-1] + upper[:-1]
def main():
    data = sys.stdin.read().split(); pos = 0
    T = int(data[pos]); pos += 1; out = []
    for _ in range(T):
        n = int(data[pos]); pos += 1
        pts = []
        for _ in range(n):
            x = int(data[pos]); y = int(data[pos+1]); pos += 2; pts.append((x, y))
        h = hull(pts)
        out.append(str(len(h)))
        for x, y in h: out.append(f'{x} {y}')
    sys.stdout.write('\n'.join(out) + '\n')
main()

File: evals/refs/hull.rs
use std::io::{self, Read, Write};

fn cross(o: (i64, i64), a: (i64, i64), b: (i64, i64)) -> i64 {
    (a.0 - o.0) * (b.1 - o.1) - (a.1 - o.1) * (b.0 - o.0)
}

fn hull(mut pts: Vec<(i64, i64)>) -> Vec<(i64, i64)> {
    pts.sort();
    pts.dedup();
    if pts.len() <= 1 {
        return pts;
    }

    let mut lower: Vec<(i64, i64)> = Vec::new();
    for &p in pts.iter() {
        while lower.len() >= 2
            && cross(lower[lower.len() - 2], lower[lower.len() - 1], p) <= 0
        {
            lower.pop();
        }
        lower.push(p);
    }
    let mut upper: Vec<(i64, i64)> = Vec::new();
    for &p in pts.iter().rev() {
        while upper.len() >= 2
            && cross(upper[upper.len() - 2], upper[upper.len() - 1], p) <= 0
        {
            upper.pop();
        }
        upper.push(p);
    }
    lower.pop();
    upper.pop();
    lower.extend(upper);
    lower
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_ascii_whitespace().map(|s| s.parse::<i64>().unwrap());

    let stdout = io::stdout();
    let mut out = io::BufWriter::new(stdout.lock());

    let t = it.next().unwrap();
    for _ in 0..t {
        let n = it.next().unwrap();
        let mut pts: Vec<(i64, i64)> = Vec::with_capacity(n as usize);
        for _ in 0..n {
            let x = it.next().unwrap();
            let y = it.next().unwrap();
            pts.push((x, y));
        }
        let h = hull(pts);
        writeln!(out, "{}", h.len()).unwrap();
        for p in &h {
            writeln!(out, "{} {}", p.0, p.1).unwrap();
        }
    }
}
